piquasso/core/_expressions.py

```diff
@@ -0,0 +1,185 @@
+#
+# Copyright 2021-2025 Budapest Quantum Computing Group
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import ast
+import operator as op
+from typing import Tuple, Any
+
+from piquasso.api.exceptions import InvalidExpression
+
+BINOPS = {
+    ast.Add: op.add,
+    ast.Sub: op.sub,
+    ast.Mult: op.mul,
+    ast.Div: op.truediv,
+    ast.Mod: op.mod,
+    ast.Pow: op.pow,
+}
+
+UNARYOPS = {
+    ast.UAdd: op.pos,
+    ast.USub: op.neg,
+    ast.Not: op.not_,
+}
+
+BOOLOPS = {
+    ast.And: all,
+    ast.Or: any,
+}
+
+CMPOPS = {
+    ast.Eq: op.eq,
+    ast.NotEq: op.ne,
+    ast.Lt: op.lt,
+    ast.LtE: op.le,
+    ast.Gt: op.gt,
+    ast.GtE: op.ge,
+}
+
+# derive ALLOWED from mappings + structural nodes
+ALLOWED = (
+    {
+        ast.Expression,
+        ast.BoolOp,
+        ast.UnaryOp,
+        ast.BinOp,
+        ast.Compare,
+        ast.Name,
+        ast.Load,
+        ast.Subscript,
+        ast.Slice,
+        ast.Constant,
+        ast.List,
+        ast.Tuple,
+        # optional: for Python < 3.9 subscripts
+        getattr(ast, "Index", object()),
+    }
+    | set(BINOPS.keys())
+    | set(UNARYOPS.keys())
+    | set(BOOLOPS.keys())
+    | set(CMPOPS.keys())
+)
+# remove the placeholder object if present
+ALLOWED.discard(object())
+
+
+class Expression:
+    """Safe expression over a single variable `x` (tuple-like)."""
+
+    def __init__(self, src: str):
+        self._src = src.strip()
+        try:
+            self._tree = ast.parse(self._src, mode="eval")
+        except SyntaxError as e:
+            raise InvalidExpression(f"Invalid syntax: {e}") from e
+        self._validate(self._tree)
+
+    def __call__(self, x: Tuple) -> Any:
```

Comment:
Why isn't `x` optional? One should be able to have expressions which don't take arguments.